Massive code clean up

rikka0w0 opened the pull request:
1. Use tabs for indentation
2. Add comments and deprecate less useful methods in FabricShieldUtils
3. Move test code from classes `FabricShieldLib` and `FabricShieldLibClient` to their own files
4. Add `FabricShieldLibClient::FABRIC_BANNER_SHIELD_MODEL_TYPE` to `FabricShieldLibClient` so that modders can avoid hardcoding that string if they prefer data generators.


I also raised a few questions regarding the `FabricShieldLib.MODEL_COMPONENT`:
1. It is not marked as final, so modders could accidentally change its value. My suggestion is to make it private and add a getter method.
2. Why did we introduce `FabricShieldModelComponent` in the first place? `FabricShieldModelComponent` only carries client-specific info, but it will be synced to the server side, which does not sound right to me.

In addition, we should highlight which classes/methods are supposed to be used by modders and kept binary compatible.

CringeStar commented on src/main/java/com/github/crimsondawn45/fabricshieldlib/initializers/FabricShieldLibClient.java:
```diff
@@ -18,157 +14,153 @@
+		SpriteAtlasTexture.BLOCK_ATLAS_TEXTURE,
+		Identifier.of(FabricShieldLib.MOD_ID, "entity/fabric_banner_shield_base")
+	);
+
```
Could you potentially move these two to the test files?

rikka0w0 replied:
Sounds better to me if we put the test code into its own package

CringeStar replied:
Yep, sounds good to me

CringeStar reviewed (changes requested):
Just the one comment, but otherwise, thanks for this, we def need to sort out some of these old structures!

CringeStar commented:
> 1. It is not marked as final, so modders could accidentally change its value. My suggestion is to make it private and add a getter method.

Yeah that's something I def missed. Pretty sure we can just make it final since lib users never actually interact with the one in FabricShieldLib.java and we never modify it either. The user only uses the record version that is its own file.

> 2. Why did we introduce `FabricShieldModelComponent` in the first place? `FabricShieldModelComponent` only carries client-specific info, but it will be synced to the server side, which does not sound right to me.

We had to introduce it to provide our model information to the renderer. Before we could mixin into the rendering process and just plop our own render method in with the model info as parameters, but with the new custom renderers and special model types, the texture info is more hardcoded, so we had to kind of sneak in our model info through a component. At least I haven't found another way to render banner shields without this process.

rikka0w0 commented:
> I feel like we can honestly just go back to having a boolean in the class that gets set in the constructor. Seems like the simplest implementation and we know it works.

I think we can get rid of `interface FabricShield` completely and replace `boolean FabricShield::displayTooltip()` and `boolean FabricShield::supportsBanner()` with the tag `fabricshieldlib:shield_show_tooltip` and `fabricshieldlib:shield_supports_banner`.

rikka0w0 commented:
> If we want to have test items that use the tag, we have to remove them before every release and re-add every time we dev again.

We can simply exclude the test tag file (`data/fabricshieldlib/tags/item/supports_banner.json`) from the release jar and will not be in the normal game.

CringeStar commented:
> We can strip the test items from tag files when creating the release jar so that they will not be in the normal game.


oh nice! I didn't know we could. Thatd also be useful for durability.json and c:shields.

Also though, I don't think the tooltip option should be via tag. I think we should make it either not an option (always true) since we already have the config for enable tooltip, or it should be in the constructor, so that people don't have to go out of their way to make tooltips visible for their shield.

CringeStar reviewed (changes requested):
Almost done, just a couple things from the convo:

- Moving the test stuff to its own package, and the spriteidentifiers in ClientInit to tests
- Re-add  modelFabricShield, but commented out for now
- Add test items to durability.json and c:shields with code to strip them too
- Change show_tooltip tag to one of the other things suggested: always true or constructor boolean

rikka0w0 commented:
> I think we should make it either not an option (always true) since we already have the config for enable tooltip

The current logic (since we introduce tooltip) is that the tooltip will always be visible for all shield items except:
1. Globally disabled in the config
2. Explicitly disabled by returning false in `FabricShield::displayTooltip`, only possible with modded shields that use FabricShieldLib.

So I think we can keep the `FabricShield` for now with `displayTooltip` as the only method prototype.

Also, should we consider mixin into vanilla `FabricShieldDecoratorRecipe` rather than having our own `FabricShieldDecoratorRecipe`? I saw that most of the codes in `FabricShieldDecoratorRecipe` are duplicates of `FabricShieldDecoratorRecipe`.

CringeStar commented:
> The current logic (since we introduce tooltip) is that the tooltip will always be visible for all shield items except:
> 
> 1. Globally disabled in the config
> 2. Explicitly disabled by returning false in `FabricShield::displayTooltip`, only possible with modded shields that use FabricShieldLib.

We could js move the displayTooltip thing into the FabricShieldItem class with the same functionality, default true. Im thinking maybe adding a component called DISABLE_TOOLTIP so that its not in the way in the constructor but the user doesn't have to go out of their way by adding it to a tag or extending the class to override.

> So I think we can keep the `FabricShield` for now with `displayTooltip` as the only method prototype.

Nah honestly, the interface became obsolete once the rendering system changed, because we've been able to distinguish between our two item types without separate classes, so we can js delete the interface.

> Also, should we consider mixin into vanilla `FabricShieldDecoratorRecipe` rather than having our own `FabricShieldDecoratorRecipe`? I saw that most of the codes in `FabricShieldDecoratorRecipe` are duplicates of `FabricShieldDecoratorRecipe`. 

Yeah that honestly makes a lot of sense, I js didn't do it originally because I was unfamiliar with the complicated mixins when you edit conditional values. We js need to modify lines 40 and 71 with our method `supportBannerAndModded(ItemStack itemStack)`. I'd probably be a WrapOperation just like the `HeldItemRendererMixin`

CringeStar commented on src/main/java/com/github/crimsondawn45/fabricshieldlib/tests/FabricShieldLibClientTests.java:
```diff
@@ -0,0 +1,67 @@
+				FabricShieldLibClientTests.FABRIC_BANNER_SHIELD_MODEL_LAYER
+			).codec;
+
+		SpecialModelTypes.ID_MAPPER.put(FABRIC_BANNER_SHIELD_MODEL_TYPE, codec);
```
Oh wait, through this new method, will library users have to make their own special model types? I feel like that's really inconvenient and unintuitive. the SpecialModelType "fabric_banner_shield" was supposed to be the type for all shields that supports banners, not specifically the test shield named that. If removing the model component requires this implementation, I feel like we should stick with the component then. @rikka0w0

CringeStar replied:
The only issue I have with this implementation is that every shield that is made will have its own model type. which sounds like a scaling issue to me. The user end is definitely less code, but I'm gonna double check with the Fabric discord to see if having many model types would be negative for performance/if there's a benefit to running all the shields through one model type/renderer.

CringeStar replied:
Tagged you, but the convo was kinda short? I got the vibe that one renderer was better but look over the convo yourself. Its in mod-dev-rendering

CringeStar replied:
Yeah if I'm being honest I didn't fully understand the wording for that solution. For the codec you could look at the deleted component class.

rikka0w0 replied:
Check out my latest commit. I believe it is the most elegant implementation. The banner shield textures are now specified in item model JSON files. Now we can use one special renderer for all modded shields.

rikka0w0 commented on src/main/java/com/github/crimsondawn45/fabricshieldlib/lib/object/FabricShieldUtils.java:
```diff
@@ -16,40 +15,100 @@
+	 */
+	@SuppressWarnings("deprecation")
+	public static boolean displayTooltip(Item item) {
+		return item.getRegistryEntry().isIn(FabricShieldTags.SHOW_TOOLTIP);
```
removed

CringeStar commented on src/main/java/com/github/crimsondawn45/fabricshieldlib/lib/object/FabricShieldModelRenderer.java:
```diff
@@ -30,78 +32,85 @@
+		public static final MapCodec<FabricShieldModelRenderer.Unbaked> CODEC = RecordCodecBuilder.mapCodec(
+			instance -> instance.group(
+					Identifier.CODEC.fieldOf("texture_banner").forGetter(FabricShieldModelRenderer.Unbaked::baseModel),
+					Identifier.CODEC.fieldOf("texture_default").forGetter(FabricShieldModelRenderer.Unbaked::baseModel)
```
Shouldnt one of these be the nopat object? Like after Unbaked::

CringeStar commented on src/main/java/com/github/crimsondawn45/fabricshieldlib/mixin/RendererMixin.java:
Im wondering if we need this still?

CringeStar replied:
or if it will have to  be changed with the new way we use the renderer. Maybe we could comment out like the model setting in client tests, since it seems like its only used for that functionality which is not yet supported?

rikka0w0 commented:
> * Recipe Mixin (I can handle this)

I will not touch the recipe Mixin in this PR, except for the formatting

> FabricShield interface -> Deprecate or remove?

My suggestion is to remove `FabricShield` interface, as we have introduced so many breaking changes in the API. We can also have a tag named `no_tooltip` as a replacement to `FabricShield::displayTooltip`. In most case, the user doesn't need to add shields to this tag. The only case required is that the user wants to explicitly disable tooltip for a given shield.

> Figure out if we currently need RendererMixin and ShieldModelSetCallback.

We don't need either at the moment, my suggestion is to deprecate `ShieldModelSetCallback`. If we want to support custom shield shape in the future, we can register our own `SpecialModelRenderer` to handle it or add the logic to `FabricShieldModelRenderer`. 

> Typo in FabricShieldModelRenderer
Fixed.

CringeStar commented:
Also maybe we wanna look at using ComponentTooltipAppenderRegistry for our tooltips? Just saw someone mention it on discord.

rikka0w0 commented:
> Also maybe we wanna look at using ComponentTooltipAppenderRegistry for our tooltips?

`ComponentTooltipAppenderRegistry` looks better to me. It has `addBefore` and `addAfter` functions, which allow us to specify some existing tooltip as the anchor without having to hack it. Our current implementation may not work as expected if the lang is set to non-English.
